adr/ADR-1.md

```diff
@@ -4,152 +4,280 @@
-indicates the specific reason for the 404 - here `10059` means the stream did not exist, helping developers identify the
-real underlying cause. 
+## Schemas
 
-More information about the `error_code` system can be found in [ADR-7](0007-error-codes.md).
+All requests and responses that are in JSON format have JSON Schema Draft 7 documents describing them.
 
-Ideally the error response includes a minimally valid body of what was requested but this can be very hard to implement correctly.
+For example the request to create a stream has the type `io.nats.jetstream.api.v1.stream_create_request` and the response
+is a `io.nats.jetstream.api.v1.stream_create_response`. We have additional Schemas that describe some subsets of information
+for example `io.nats.jetstream.api.v1.stream_configuration` describes a valid Stream Configuration.
 
-Today the list API's just return `["ORDERS"]`, these will become:
+Generally requests do not expect the kind - it's inferred from which subject is accessed - but replies all have type hints.
 
-```json
+The `nats` CLI can list, view and validate documents against these schemas:
+
+```nohighlight
+$ nats schema list stream_create
+Matched Schemas:
+
+  io.nats.jetstream.api.v1.stream_create_request
+  io.nats.jetstream.api.v1.stream_create_response
+```
+
+The document can be viewed, pass `--yaml` to view it in YAML format that's easier for reading:
+
+```nohighlight
+$ nats schema show io.nats.jetstream.api.v1.stream_create_request
 {
-  "type": "io.nats.jetstream.api.v1.stream_list",
-  "time": "2020-04-23T16:51:18.516363Z",
-  "streams": [
-    "ORDERS"
-  ]
-}
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "$id": "https://nats.io/schemas/jetstream/api/v1/stream_create_request.json",
+  "description": "A request to the JetStream $JS.API.STREAM.CREATE API",
+...
+```
+
+Finally, if you're developing an agent it could be useful to capture the JSON documents and validate them, the CLI can
+help with that:
+
+```nohighlight
+$ nats schema validate io.nats.jetstream.api.v1.stream_create_request x.json
+Validation errors in x.json:
+
+  (root): max_consumers is required
+  (root): max_msgs is required
 ```
 
-With the same `error` treatment when some error happens.
+Here we validated the `x.json` file and found a number of errors.
 
-## Implementation
+If you're writing a client and have JSON Schema validators to hand you can access the schemas in our [schema repository](https://github.com/nats-io/jsm.go/tree/main/schemas)
 
-While implementing this in JetStream the following pattern emerged:
+## Data Types
 
-```go
-type JSApiResponse struct {
-	Type  string    `json:"type"`
-	Error *ApiError `json:"error,omitempty"`
+The JetStream API has many data types, generally the JSON Schema tries to point this out, though we have some gaps in the data for sure.
+This section explore a few of the types in details.
```

Comment:
I'll add a bit here - since the schemas will show different minimums for example many signed int64s have a minimum value of -1 for example in the schema, clients must do the right thing there.